testinfra/utils/ansible_runner.py

```diff
@@ -18,150 +18,183 @@
+
+        env = os.environ.copy()
+        env['ANSIBLE_NOCOLOR'] = "1"
+        p = subprocess.Popen(
+            cmd,
+            shell=True,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env=env
+        )
+        (so, se) = p.communicate()
 
+        if p.returncode != 0:
+            msg = 'ansible-inventory failed: %s' % se
+            raise AnsibleInventoryException(msg)
 
-class Callback(ansible.plugins.callback.CallbackBase):
+        inv = json.loads(so)
 
-    def __init__(self, *args, **kwargs):
-        self.result = {}
-        super(Callback, self).__init__(*args, **kwargs)
+        return inv
 
-    def runner_on_ok(self, host, result):
-        self.result = result
+    def get_hosts(self, pattern=None):
+        '''Return a list of host names from inventory via the pattern'''
+        inv = self.fetch_inventory()
+        return list(inv['_meta']['hostvars'].keys())
 
-    def runner_on_failed(self, host, result, ignore_errors=False):
-        self.result = result
+    def get_variables(self, host, refresh=True):
+        '''Get a mixture of inventory vars and magic vars'''
 
-    # pylint: disable=no-self-use
-    def runner_on_unreachable(self, host, result):
-        raise RuntimeError(
-            'Host {} is unreachable: {}'.format(
-                host, pprint.pformat(result)),
+        if host not in self.variables or refresh:
+
+            _vars = self.variables.get(host, {})
+
+            # inventory vars
+            _vars.update(self.fetch_inventory(host=host))
+
+            # this is a hack to get the magic vars
+            res = self.run(host, 'debug', 'var=hostvars')
+            _vars.update(res.get('hostvars', {}).get(host, {}))
+
+            # one of the unit tests insist this should be returned
+            _vars['inventory_hostname'] = host
+            self.variables[host] = _vars
+
+        return self.variables[host]
+
+    def run(self, host, module_name, module_args=None, **kwargs):
+        '''Invokes a single module on a single host and returns dict results'''
+
+        # runner must have a directory based payload
+        data_dir = tempfile.mkdtemp(prefix='runner.data.%s.' \
+            % datetime.datetime.now().isoformat().replace(':', '-'))
+        if os.path.exists(data_dir):
+            shutil.rmtree(data_dir)
+        if not os.path.exists(data_dir):
+            os.makedirs(data_dir)
+
+        # runner must have an inventory file
+        inv_dir = os.path.join(data_dir, 'inventory')
+        if not os.path.exists(inv_dir):
+            os.makedirs(inv_dir)
+        shutil.copy(
+            self.host_list,
+            os.path.join(inv_dir, os.path.basename(self.host_list))
         )
 
-    def runner_on_skipped(self, host, item=None):
-        self.result = {
-            'failed': True,
-            'msg': 'Skipped. You might want to try check=False',
-            'item': item,
+        # molecule inventories use lookups
```

Comment:
Can you explain what is this ``envvars`` file about ?

Reply:
Sure ... so molecule exports a lot of environment variables which are picked up by the yaml based inventory it wants to give to ansible. Here's some examples ...

```
MOLECULE_DEBUG: 'False'
MOLECULE_DEPENDENCY_NAME: galaxy
MOLECULE_DRIVER_NAME: docker
MOLECULE_ENV_FILE: /home/jtanner/workspace/github/jctanner/2019-04-16-ecosystem/foorole/.env.yml
MOLECULE_EPHEMERAL_DIRECTORY: /tmp/molecule/foorole/default
MOLECULE_FILE: /home/jtanner/workspace/github/jctanner/2019-04-16-ecosystem/foorole/molecule/default/molecule.yml
MOLECULE_INSTANCE_CONFIG: /tmp/molecule/foorole/default/instance_config.yml
MOLECULE_INVENTORY_FILE: /tmp/molecule/foorole/default/inventory/ansible_inventory.yml
MOLECULE_LINT_NAME: yamllint
MOLECULE_PROJECT_DIRECTORY: /home/jtanner/workspace/github/jctanner/2019-04-16-ecosystem/foorole
MOLECULE_PROVISIONER_LINT_NAME: ansible-lint
MOLECULE_PROVISIONER_NAME: ansible
MOLECULE_SCENARIO_DIRECTORY: /home/jtanner/workspace/github/jctanner/2019-04-16-ecosystem/foorole/molecule/default
MOLECULE_SCENARIO_NAME: default
MOLECULE_VERIFIER_LINT_NAME: flake8
MOLECULE_VERIFIER_NAME: testinfra
MOLECULE_VERIFIER_TEST_DIRECTORY: /home/jtanner/workspace/github/jctanner/2019-04-16-ecosystem/foorole/molecule/default/tests
```

Here's a sample inventory ...
```
---
all:
  hosts:
    instance: &id001
      ansible_connection: docker
  vars:
    molecule_ephemeral_directory: '{{ lookup(''env'', ''MOLECULE_EPHEMERAL_DIRECTORY'')
      }}'
    molecule_file: '{{ lookup(''env'', ''MOLECULE_FILE'') }}'
    molecule_instance_config: '{{ lookup(''env'', ''MOLECULE_INSTANCE_CONFIG'') }}'
    molecule_no_log: '{{ lookup(''env'', ''MOLECULE_NO_LOG'') or not molecule_yml.provisioner.log|default(False)
      | bool }}'
    molecule_scenario_directory: '{{ lookup(''env'', ''MOLECULE_SCENARIO_DIRECTORY'')
      }}'
    molecule_yml: '{{ lookup(''file'', molecule_file) | molecule_from_yaml }}'
ungrouped:
  hosts:
    instance: *id001
  vars: {}
```

Ansible-runner makes heavy use of setting environment variables in the pexect spawn environment it creates, so users have to supply any custom env vars they care about. Since the vars molecule exports won't make it down through pytest and then testinfra and then runner and then pexect, I have to tell runner about those vars and their values.